File: 112.path-sum.java
/*
 * @lc app=leetcode id=112 lang=java
 *
 * [112] Path Sum
 */

// @lc code=start

import java.util.Stack;
import java.util.AbstractMap.SimpleEntry;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        if (root == null)
            return false;
        Stack<SimpleEntry<TreeNode, Integer>> stack = new Stack<>();
        stack.push(new SimpleEntry<>(root, targetSum));
        boolean ret = false;
        while (!stack.empty()) {
            SimpleEntry<TreeNode, Integer> entry = stack.pop();
            if (entry != null) {
                TreeNode node = entry.getKey();
                int target = entry.getValue();
                if (node.left != null) {
                    stack.push(new SimpleEntry<>(node.left, target - node.val));
                }
                if (node.right != null) {
                    stack.push(new SimpleEntry<>(node.right, target - node.val));
                }
                stack.push(new SimpleEntry<>(node, target));
                stack.push(null);
            } else {
                entry = stack.pop();
                TreeNode node = entry.getKey();
                int target = entry.getValue();
                if (node.left == null && node.right == null && node.val == target)
                    ret = true;
            }
        }
        return ret;
    }
}
// @lc code=end

File: 112.path-sum.cpp
/*
 * @lc app=leetcode id=112 lang=cpp
 *
 * [112] Path Sum
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

#include <stack>

using namespace std;

class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        if (root == nullptr) return false;
        stack<TreeNode *> st;
        stack<int> presum;
        st.push(root);
        presum.push(0);
        while (!st.empty()) {
            TreeNode *node = st.top(); st.pop();
            int sum = presum.top(); presum.pop();
            if (node != nullptr) {
                if (node->right != nullptr) {
                    st.push(node->right);
                    presum.push(sum + node->val);
                }
                if (node->left != nullptr) {
                    st.push(node->left);
                    presum.push(sum + node->val);
                }
                st.push(node); st.push(nullptr); presum.push(sum);
            } else {
                node = st.top(); st.pop();
                if (node->left == nullptr && node->right == nullptr) {
                    if (sum + node->val == targetSum) return true;
                }
            }
        }
        return false;
    }
};
// @lc code=end
